arangod/Aql/CollectionAccess.h

```diff
@@ -0,0 +1,97 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGOD_AQL_COLLECTIONACCESS_H
+#define ARANGOD_AQL_COLLECTIONACCESS_H
+
+#include <memory>
+#include <optional>
+#include <unordered_map>
+
+#include "Aql/ExecutionNodeId.h"
+
+namespace arangodb::velocypack {
+class Slice;
+}
+
+namespace arangodb::aql {
+class Collections;
+class ExecutionNode;
+struct Collection;
+struct Variable;
+
+/**
+ * @brief Holds information how an aql::Collection is accessed. This is on the
+ * Node level during query planning, so can be different for different accesses
+ * of the same collection. It's not used to differentiate shards.
+ */
+class CollectionAccess {
+ public:
+  CollectionAccess() = default;
+  explicit CollectionAccess(aql::Collection const*);
+  CollectionAccess(aql::Collections const*, velocypack::Slice);
+
+  auto collection() const noexcept -> aql::Collection const*;
+
+  /// @brief modify collection after cloning
+  /// should be used only in smart-graph context!
```

Comment:
Hm, would then maybe sense to rename the function? (So that everyone knows that this should only be used within the smart-graph context?)

Reply:
```
  void setCollection(aql::Collection const* collection);

```

Reply:
I actually just copy-pasted that comment when delegating this from `CollectionAccessingNode` to `CollectionAccess`. I've got only a hunch what this comment is about, would have to research first, and that seems too unrelated to me for this PR.